[23674] Migrate wiki titles to URL slugs

oliverguenther opened the pull request:
OpenProject < 6.0.0 processed wiki titles in a specific fashion (#titleize)
replacing spaces with underscores and removing some characters.
With OpenProject 6.0.0, titles were allowed to contain any chars.
This causes issues since a title `foo bar` was saved as `foo_bar`, but displayed with spaces again
and in turn, many page links actualy use [[foo bar]] instead of the actual title.

Spaces were no longer matching the existing pages.

Additionally, not all characters can be properly detected in the `parse_wiki_links` steps and this makes it impossible to link to some pages. (e.g.,`[[Foo & Bar]]`)

To remove this issue, we introduce wiki slugs using `stringex` to create proper permalinks that can be linked to, and only fall back to finding by titles for the old pages.

https://community.openproject.com/work_packages/23674

oliverguenther committed:
Migrate wiki titles to URL slugs

Wiki titles in 6.0 can contain ANY character, which is causing issues
for specific characters processed by textile and CGI-escaped before we
can actually parse wiki links.

To mitigate the issue of linking space/underscored wiki pages, introduce
a permalink URL based on the title and create a migration to move old
underscored titles
(e.g., `base_de_donées` to spaced titles `base de donées`).

oliverguenther committed:
Remove `pretty_title` since that won't matter now.

oliverguenther committed:
For new pages, use the title itself as a slug input

oliverguenther committed:
Let text_formatter write the slug as a link

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://oliverguenther-fix-23674-wiki-titles.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/' target='_blank'>Open stage</a> or <a href='http://oliverguenther-fix-23674-wiki-titles.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com//teatro.log' target='_blank'>view logs</a>.

machisuji commented on app/views/wiki/destroy.html.erb:
```diff
@@ -26,8 +26,8 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
-<%= toolbar title: @page.pretty_title %>
-<% html_title l(:button_delete), @page.pretty_title %>
+<%= toolbar title: @page.title %>
+<% html_title l(:button_delete), @page.title %>
```
@oliverguenther Could you use `t` instead of `l` here while you're at it, please?

machisuji commented on app/views/wiki/annotate.html.erb:
```diff
@@ -27,7 +27,7 @@ See doc/COPYRIGHT.rdoc for more details.
 ++#%>
 
-<% html_title l(:project_module_wiki), l(:label_history), l(:text_comment_wiki_page, page: @page.pretty_title)  %>
+<% html_title l(:project_module_wiki), l(:label_history), l(:text_comment_wiki_page, page: @page.title)  %>
```
Here too. <3

oliverguenther replied:
done!

machisuji commented on db/migrate/20160726090624_add_slug_to_wiki_pages.rb:
```diff
@@ -0,0 +1,61 @@
+  def down
+    remove_column :wiki_pages, :slug
+
+    # Restore the old title
```
It would make sense to put this into a method as well (`rollback_titles`).

oliverguenther replied:
done!

machisuji commented on db/migrate/20160726090624_add_slug_to_wiki_pages.rb:
```diff
@@ -0,0 +1,61 @@
+    add_index :wiki_pages, [:wiki_id, :slug], name: 'wiki_pages_wiki_id_slug', unique: true
+
+    migrate_titles
+    change_column_null :wiki_pages, :slug, true
```
Why can't we allow this right away in line 3?
Even if there is a good reason for it me asking this is a good sign that a comment would be helpful.

oliverguenther replied:
I've added a comment on why it's necessary. In line 3, wiki_pages still contain NULL values for the slug, and this would raise an error when trying to add a `null: false` column.

machisuji commented on spec/features/wiki/wiki_unicode_spec.rb:
```diff
@@ -0,0 +1,113 @@
+
+    [[base-de-donnees]] should link to wiki_page_2
+
+    [[base-de-donnees-1]] should link to wiki_page_3
```
How does the base-de-donnees**-1** come about?

oliverguenther replied:
This is due to the same title being used in wiki_page2 and wiki_page3. The page is saved correctly, but the `acts_as_url` takes care of resolving conflicting slugs.

I've changed the spec to make clear why that suffix is added.

machisuji replied:
Ah I see cool.

oliverguenther committed:
Scope unique slugs per wiki_id

oliverguenther committed:
Fix legacy specs

oliverguenther committed:
Fix lookup for menu-items controller

oliverguenther committed:
Fix cukes

oliverguenther committed:
Fix specs

oliverguenther committed:
Replace l with t

oliverguenther committed:
Clarify slug duplicate in spec

oliverguenther committed:
Add separate method for downgrade titles

oliverguenther committed:
Fix null value in migration

machisuji commented on app/views/wiki/edit.html.erb:
```diff
@@ -26,8 +26,8 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
-<%= toolbar title: @page.pretty_title %>
-<% html_title l(:label_edit), @page.pretty_title %>
+<%= toolbar title: @page.title %>
+<% html_title l(:label_edit), @page.title %>
```
`t` please

oliverguenther committed:
Titles no longer need to be unique

Since the slug is now the unique identifier for the page, we can
avoid title conflicts by letting the slug determine unique records.

machisuji commented:
👍
